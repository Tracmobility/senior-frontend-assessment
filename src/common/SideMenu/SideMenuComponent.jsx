import React, { useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faLaptop, faProjectDiagram, faMap, faListAlt, faPlusSquare, faChevronRight, faCogs,
  faQrcode, faChevronDown
} from '@fortawesome/free-solid-svg-icons';
import styles from './SideMenuComponent.module.scss';

const SideMenuComponent = () => {
  const [operationMenuDropdown, setOperationMenuDropdown] = useState(true);

  const handleOperationMenuDropdown = () => {
    if (operationMenuDropdown === true) {
      setOperationMenuDropdown(false);
    } else {
      setOperationMenuDropdown(true);
    }
    return true;
  }

  return (
    <nav className={styles.sideMenuContainer}>
      <div className={styles.tracBrand}>
        <FontAwesomeIcon icon={faLaptop}/>
        <h1>TRACMOBILITY</h1>
      </div>
      <div className={styles.kpiSection}>
        <FontAwesomeIcon icon={faProjectDiagram}/>
        <h3>KPI Dashboard</h3>
      </div>
      <div className={styles.taskManagementSection}>
        <h3>Task Management</h3>
        <a href="#">
          <FontAwesomeIcon icon={faMap}/>
          <p>Task Map</p>
        </a>
        <a href="#">
          <FontAwesomeIcon icon={faListAlt}/>
          <p>Task List</p>
        </a>
        <a href="#">
          <FontAwesomeIcon icon={faPlusSquare}/>
          <p>Task Creation</p>
          <FontAwesomeIcon icon={faChevronRight} className={styles.taskCreationChevron} />
        </a>
        <a href="#">
          <FontAwesomeIcon icon={faCogs} />
          <p>Rule Setting</p>
        </a>
      </div>
      <div className={styles.operationCenterSection}>
        <h3>Operation Center</h3>
        <div className={styles.operationMenuContainer}>
          {
            operationMenuDropdown ? (
              <>
                <div className={styles.operationButtonContainer}>
                  <FontAwesomeIcon icon={faQrcode}/>
                  <button onClick={handleOperationMenuDropdown} className={styles.operationDropdown}>Vehicle Information</button>
                  <FontAwesomeIcon icon={faChevronDown} />
                </div>
                <div className={styles.operationMenuDropdown}>
                  <a href="/">Vehicle List</a>
                  <a href="/vehicle-map">Vehicle Map</a>
                  <a href="#">Vehicle Search</a>
                </div>
              </>
            ) : (
              <div className={styles.operationButtonContainer}>
                <FontAwesomeIcon icon={faQrcode}/>
                <button onClick={handleOperationMenuDropdown} className={styles.operationDropdown}>Vehicle Information</button>
                <FontAwesomeIcon icon={faChevronRight} />
              </div>
            )
          }
        </div>
      </div>
    </nav>
  );
};

export default SideMenuComponent;
