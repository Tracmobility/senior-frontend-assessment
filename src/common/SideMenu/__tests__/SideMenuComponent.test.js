import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import SideMenuComponent from '../SideMenuComponent';

Enzyme.configure({ adapter: new Adapter() });

const sideMenuComponent = (<SideMenuComponent />);

describe('The appearance and functionality of the side menu component', () => {
  const container = mount(sideMenuComponent);

  it('Matches the SideMenuComponent snapshot', () => {
    const tree = renderer.create(sideMenuComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Contains the master nav container with the correct attributes', () => {
    expect(container.childAt(0).hasClass('sideMenuContainer')).toBe(true);
    expect(container.childAt(0).name()).toBe('nav');
  });

  it('Displays the Tracmobility branding correctly', () => {
    expect(container.childAt(0).childAt(0).hasClass('tracBrand')).toBe(true);
    expect(container.find('.tracBrand').childAt(0).name()).toBe('FontAwesomeIcon');
    expect(container.find('.tracBrand').childAt(1).text()).toBe('TRACMOBILITY');
  });

  it('Contains all of the empty links for other functionality', () => {
    const taskManagementSection = container.find('.taskManagementSection');
    expect(taskManagementSection.childAt(1).prop('href')).toBe('#');
    expect(taskManagementSection.childAt(2).prop('href')).toBe('#');
    expect(taskManagementSection.childAt(3).prop('href')).toBe('#');
    expect(taskManagementSection.childAt(4).prop('href')).toBe('#');
  });

  it('Displays the Operation Menu Dropdown to contain working links for navigation', () => {
    const operationMenuDropdown = container.find('.operationMenuDropdown');
    expect(operationMenuDropdown.childAt(0).prop('href')).toBe('/');
    expect(operationMenuDropdown.childAt(1).prop('href')).toBe('/vehicle-map')
  })
})