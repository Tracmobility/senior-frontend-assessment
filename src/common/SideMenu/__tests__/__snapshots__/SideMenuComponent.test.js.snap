// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The appearance and functionality of the side menu component Matches the SideMenuComponent snapshot 1`] = `
<nav
  className="sideMenuContainer"
>
  <div
    className="tracBrand"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-laptop fa-w-20 "
      data-icon="laptop"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 640 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M624 416H381.54c-.74 19.81-14.71 32-32.74 32H288c-18.69 0-33.02-17.47-32.77-32H16c-8.8 0-16 7.2-16 16v16c0 35.2 28.8 64 64 64h512c35.2 0 64-28.8 64-64v-16c0-8.8-7.2-16-16-16zM576 48c0-26.4-21.6-48-48-48H112C85.6 0 64 21.6 64 48v336h512V48zm-64 272H128V64h384v256z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <h1>
      TRACMOBILITY
    </h1>
  </div>
  <div
    className="kpiSection"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-project-diagram fa-w-20 "
      data-icon="project-diagram"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 640 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M384 320H256c-17.67 0-32 14.33-32 32v128c0 17.67 14.33 32 32 32h128c17.67 0 32-14.33 32-32V352c0-17.67-14.33-32-32-32zM192 32c0-17.67-14.33-32-32-32H32C14.33 0 0 14.33 0 32v128c0 17.67 14.33 32 32 32h95.72l73.16 128.04C211.98 300.98 232.4 288 256 288h.28L192 175.51V128h224V64H192V32zM608 0H480c-17.67 0-32 14.33-32 32v128c0 17.67 14.33 32 32 32h128c17.67 0 32-14.33 32-32V32c0-17.67-14.33-32-32-32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <h3>
      KPI Dashboard
    </h3>
  </div>
  <div
    className="taskManagementSection"
  >
    <h3>
      Task Management
    </h3>
    <a
      href="#"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-map fa-w-18 "
        data-icon="map"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 117.66v346.32c0 11.32 11.43 19.06 21.94 14.86L160 416V32L20.12 87.95A32.006 32.006 0 0 0 0 117.66zM192 416l192 64V96L192 32v384zM554.06 33.16L416 96v384l139.88-55.95A31.996 31.996 0 0 0 576 394.34V48.02c0-11.32-11.43-19.06-21.94-14.86z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <p>
        Task Map
      </p>
    </a>
    <a
      href="#"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-list-alt fa-w-16 "
        data-icon="list-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M464 480H48c-26.51 0-48-21.49-48-48V80c0-26.51 21.49-48 48-48h416c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48zM128 120c-22.091 0-40 17.909-40 40s17.909 40 40 40 40-17.909 40-40-17.909-40-40-40zm0 96c-22.091 0-40 17.909-40 40s17.909 40 40 40 40-17.909 40-40-17.909-40-40-40zm0 96c-22.091 0-40 17.909-40 40s17.909 40 40 40 40-17.909 40-40-17.909-40-40-40zm288-136v-32c0-6.627-5.373-12-12-12H204c-6.627 0-12 5.373-12 12v32c0 6.627 5.373 12 12 12h200c6.627 0 12-5.373 12-12zm0 96v-32c0-6.627-5.373-12-12-12H204c-6.627 0-12 5.373-12 12v32c0 6.627 5.373 12 12 12h200c6.627 0 12-5.373 12-12zm0 96v-32c0-6.627-5.373-12-12-12H204c-6.627 0-12 5.373-12 12v32c0 6.627 5.373 12 12 12h200c6.627 0 12-5.373 12-12z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <p>
        Task List
      </p>
    </a>
    <a
      href="#"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-plus-square fa-w-14 "
        data-icon="plus-square"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <p>
        Task Creation
      </p>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-chevron-right fa-w-10 taskCreationChevron"
        data-icon="chevron-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </a>
    <a
      href="#"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-cogs fa-w-20 "
        data-icon="cogs"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512.1 191l-8.2 14.3c-3 5.3-9.4 7.5-15.1 5.4-11.8-4.4-22.6-10.7-32.1-18.6-4.6-3.8-5.8-10.5-2.8-15.7l8.2-14.3c-6.9-8-12.3-17.3-15.9-27.4h-16.5c-6 0-11.2-4.3-12.2-10.3-2-12-2.1-24.6 0-37.1 1-6 6.2-10.4 12.2-10.4h16.5c3.6-10.1 9-19.4 15.9-27.4l-8.2-14.3c-3-5.2-1.9-11.9 2.8-15.7 9.5-7.9 20.4-14.2 32.1-18.6 5.7-2.1 12.1.1 15.1 5.4l8.2 14.3c10.5-1.9 21.2-1.9 31.7 0L552 6.3c3-5.3 9.4-7.5 15.1-5.4 11.8 4.4 22.6 10.7 32.1 18.6 4.6 3.8 5.8 10.5 2.8 15.7l-8.2 14.3c6.9 8 12.3 17.3 15.9 27.4h16.5c6 0 11.2 4.3 12.2 10.3 2 12 2.1 24.6 0 37.1-1 6-6.2 10.4-12.2 10.4h-16.5c-3.6 10.1-9 19.4-15.9 27.4l8.2 14.3c3 5.2 1.9 11.9-2.8 15.7-9.5 7.9-20.4 14.2-32.1 18.6-5.7 2.1-12.1-.1-15.1-5.4l-8.2-14.3c-10.4 1.9-21.2 1.9-31.7 0zm-10.5-58.8c38.5 29.6 82.4-14.3 52.8-52.8-38.5-29.7-82.4 14.3-52.8 52.8zM386.3 286.1l33.7 16.8c10.1 5.8 14.5 18.1 10.5 29.1-8.9 24.2-26.4 46.4-42.6 65.8-7.4 8.9-20.2 11.1-30.3 5.3l-29.1-16.8c-16 13.7-34.6 24.6-54.9 31.7v33.6c0 11.6-8.3 21.6-19.7 23.6-24.6 4.2-50.4 4.4-75.9 0-11.5-2-20-11.9-20-23.6V418c-20.3-7.2-38.9-18-54.9-31.7L74 403c-10 5.8-22.9 3.6-30.3-5.3-16.2-19.4-33.3-41.6-42.2-65.7-4-10.9.4-23.2 10.5-29.1l33.3-16.8c-3.9-20.9-3.9-42.4 0-63.4L12 205.8c-10.1-5.8-14.6-18.1-10.5-29 8.9-24.2 26-46.4 42.2-65.8 7.4-8.9 20.2-11.1 30.3-5.3l29.1 16.8c16-13.7 34.6-24.6 54.9-31.7V57.1c0-11.5 8.2-21.5 19.6-23.5 24.6-4.2 50.5-4.4 76-.1 11.5 2 20 11.9 20 23.6v33.6c20.3 7.2 38.9 18 54.9 31.7l29.1-16.8c10-5.8 22.9-3.6 30.3 5.3 16.2 19.4 33.2 41.6 42.1 65.8 4 10.9.1 23.2-10 29.1l-33.7 16.8c3.9 21 3.9 42.5 0 63.5zm-117.6 21.1c59.2-77-28.7-164.9-105.7-105.7-59.2 77 28.7 164.9 105.7 105.7zm243.4 182.7l-8.2 14.3c-3 5.3-9.4 7.5-15.1 5.4-11.8-4.4-22.6-10.7-32.1-18.6-4.6-3.8-5.8-10.5-2.8-15.7l8.2-14.3c-6.9-8-12.3-17.3-15.9-27.4h-16.5c-6 0-11.2-4.3-12.2-10.3-2-12-2.1-24.6 0-37.1 1-6 6.2-10.4 12.2-10.4h16.5c3.6-10.1 9-19.4 15.9-27.4l-8.2-14.3c-3-5.2-1.9-11.9 2.8-15.7 9.5-7.9 20.4-14.2 32.1-18.6 5.7-2.1 12.1.1 15.1 5.4l8.2 14.3c10.5-1.9 21.2-1.9 31.7 0l8.2-14.3c3-5.3 9.4-7.5 15.1-5.4 11.8 4.4 22.6 10.7 32.1 18.6 4.6 3.8 5.8 10.5 2.8 15.7l-8.2 14.3c6.9 8 12.3 17.3 15.9 27.4h16.5c6 0 11.2 4.3 12.2 10.3 2 12 2.1 24.6 0 37.1-1 6-6.2 10.4-12.2 10.4h-16.5c-3.6 10.1-9 19.4-15.9 27.4l8.2 14.3c3 5.2 1.9 11.9-2.8 15.7-9.5 7.9-20.4 14.2-32.1 18.6-5.7 2.1-12.1-.1-15.1-5.4l-8.2-14.3c-10.4 1.9-21.2 1.9-31.7 0zM501.6 431c38.5 29.6 82.4-14.3 52.8-52.8-38.5-29.6-82.4 14.3-52.8 52.8z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <p>
        Rule Setting
      </p>
    </a>
  </div>
  <div
    className="operationCenterSection"
  >
    <h3>
      Operation Center
    </h3>
    <div
      className="operationMenuContainer"
    >
      <div
        className="operationButtonContainer"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-qrcode fa-w-14 "
          data-icon="qrcode"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 224h192V32H0v192zM64 96h64v64H64V96zm192-64v192h192V32H256zm128 128h-64V96h64v64zM0 480h192V288H0v192zm64-128h64v64H64v-64zm352-64h32v128h-96v-32h-32v96h-64V288h96v32h64v-32zm0 160h32v32h-32v-32zm-64 0h32v32h-32v-32z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <button
          className="operationDropdown"
          onClick={[Function]}
        >
          Vehicle Information
        </button>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-down fa-w-14 "
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
      <div
        className="operationMenuDropdown"
      >
        <a
          href="/"
        >
          Vehicle List
        </a>
        <a
          href="/vehicle-map"
        >
          Vehicle Map
        </a>
        <a
          href="#"
        >
          Vehicle Search
        </a>
      </div>
    </div>
  </div>
</nav>
`;
